src/App/Fossa/Container.hs

```diff
@@ -0,0 +1,208 @@
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE TypeApplications #-}
+
+module App.Fossa.Container
+  ( ImageText (..),
+    imageTextArg,
+    Locator (..),
+    SyftResponse (..),
+    ResponseArtifact (..),
+    ResponseSource (..),
+    ResponseDistro (..),
+    SourceTarget (..),
+    ContainerScan (..),
+    ContainerImage (..),
+    ContainerArtifact (..),
+    runSyft,
+    toContainerScan,
+    extractRevision,
+  )
+where
+
+import App.Fossa.EmbeddedBinary (BinaryPaths, toExecutablePath, withSyftBinary)
+import App.Types (ProjectRevision (..), OverrideProject (..))
+import Control.Effect.Diagnostics hiding (fromMaybe)
+import Control.Effect.Lift (Lift)
+import Control.Monad.IO.Class
+import Data.Aeson
+import qualified Data.Map.Lazy as LMap
+import Data.Map.Strict (Map)
+import Data.Maybe (listToMaybe, fromMaybe)
+import Data.Text (Text, pack)
+import Data.Text.Extra (breakOnAndRemove)
+import Effect.Exec (AllowErr (Never), Command (..), execJson, runExecIO)
+import Options.Applicative (Parser, argument, help, metavar, str)
+import Path
+
+newtype ImageText = ImageText {unImageText :: Text} deriving (Show, Eq, Ord)
+
+imageTextArg :: Parser ImageText
+imageTextArg = ImageText . pack <$> argument str (metavar "IMAGE" <> help "The image to scan")
+
+newtype Locator = Locator {unLocator :: Text} deriving (Eq, Ord, Show)
+
+-- | The output of the syft binary
+data SyftResponse
+  = SyftResponse
+      { responseArtifacts :: [ResponseArtifact],
+        responseSource :: ResponseSource,
+        responseDistro :: ResponseDistro
+      }
+
+instance FromJSON SyftResponse where
+  parseJSON = withObject "SyftResponse" $ \obj ->
+    SyftResponse <$> obj .: "artifacts"
+      <*> obj .: "source"
+      <*> obj .: "distro"
+
+data ResponseArtifact
+  = ResponseArtifact
+      { artifactName :: Text,
+        artifactVersion :: Text,
+        artifactType :: Text,
+        artifactPkgUrl :: Text,
+        artifactMetadataType :: Text,
+        artifactMetadata :: Map Text Value
+      }
+
+instance FromJSON ResponseArtifact where
+  parseJSON = withObject "ResponseArtifact" $ \obj ->
+    ResponseArtifact <$> obj .: "name"
+      <*> obj .: "version"
+      <*> obj .: "type"
+      <*> obj .: "purl"
+      <*> obj .: "metadataType"
+      -- We delete "files" as early as possible, which reduces
```

Comment:
Could you elaborate on what a file is? I get why we need a comment here but I don't really understand what it is saying.

Reply:
There's some output from the `syft` binary that is under the nested json key `artifacts[n].metadata.files`.

```json5
{
  "artifacts": [
    {
      "metadata": {
        "files": [
          // enormous array of info we don't need
        ]
      }
    }
  ]
}
```

Since this value is unnecessary AND huge, we take steps to prevent loading it into memory as much as possible.  Since haskell lazily evaluates data by default, and the `Data.Map` and `Data.Map.Lazy` internal map structure are strict in keys, but lazy in values, we can potentially remove the files key from this map without ever trying to hold the entire array in memory at once.

The comment is saying "we retain the json-like map structure, but we want to ignore the 'files' field to save a ton of memory."